From 14a8d894d052c4c2b8fe4526f6da75f1a581e2db Mon Sep 17 00:00:00 2001
From: Timothy Arceri <tarceri@itsqueeze.com>
Date: Mon, 17 Nov 2025 15:13:14 +1100
Subject: [PATCH] mesa: fix _mesa_update_texture_matrices()

_math_matrix_is_dirty() should only be used to decide if we need to
run _math_matrix_analyse(). We already decided that we had a new
texture matrix when we called _mesa_update_texture_matrices() so
we need to set _TexMatEnabled correctly otherwise we might
incorrectly return _NEW_FF_VERT_PROGRAM | _NEW_FF_FRAG_PROGRAM in
the following if-statement.

Fixes: ec978e002f59 ("mesa: only update fixed-func programs on texture matrix enablement changes")
Closes: https://gitlab.freedesktop.org/mesa/mesa/-/issues/14286
Reviewed-by: Emma Anholt <emma@anholt.net>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/38473>
(cherry picked from commit b0047be0c253d1fdaf8ba015704deb5380ce7fc3)
---
 .pick_status.json        |  2 +-
 src/mesa/main/texstate.c | 10 +++++-----
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/.pick_status.json b/.pick_status.json
index d203fd2977cab..934dc97ca015d 100644
--- a/.pick_status.json
+++ b/.pick_status.json
@@ -864,7 +864,7 @@
         "description": "mesa: fix _mesa_update_texture_matrices()",
         "nominated": true,
         "nomination_type": 2,
-        "resolution": 0,
+        "resolution": 1,
         "main_sha": null,
         "because_sha": "ec978e002f59c8ed829e354f8ca8d9440df487f1",
         "notes": null
diff --git a/src/mesa/main/texstate.c b/src/mesa/main/texstate.c
index 9917d6730ad66..b8746a2cac8b9 100644
--- a/src/mesa/main/texstate.c
+++ b/src/mesa/main/texstate.c
@@ -401,12 +401,12 @@ _mesa_update_texture_matrices(struct gl_context *ctx)
    for (u = 0; u < ctx->Const.MaxTextureCoordUnits; u++) {
       assert(u < ARRAY_SIZE(ctx->TextureMatrixStack));
       if (_math_matrix_is_dirty(ctx->TextureMatrixStack[u].Top)) {
-	 _math_matrix_analyse( ctx->TextureMatrixStack[u].Top );
-
-	 if (ctx->Texture.Unit[u]._Current &&
-	     ctx->TextureMatrixStack[u].Top->type != MATRIX_IDENTITY)
-	    ctx->Texture._TexMatEnabled |= ENABLE_TEXMAT(u);
+         _math_matrix_analyse(ctx->TextureMatrixStack[u].Top);
       }
+
+      if (ctx->Texture.Unit[u]._Current &&
+          ctx->TextureMatrixStack[u].Top->type != MATRIX_IDENTITY)
+         ctx->Texture._TexMatEnabled |= ENABLE_TEXMAT(u);
    }
 
    if (old_texmat_enabled != ctx->Texture._TexMatEnabled)
-- 
GitLab

